### ISSUE TITLE
Interact with collision event ?

### ISSUE THREAD

**Issue by evan-boissonnot** [Annoraaq/grid-engine#235]
Collision is ok in our game.

But, I want to go to other scene (start a new one) when colliding on a house (for example).
How can I detect it using grid engine ?

Is it possible ?

I see some of yours answers about position changed, but I don't like this approch. 
But maybe it's the only way ?

**Comment by zewa666** [comment]
hey there I think this also goes a bit in the direction of what I've asked here https://github.com/Annoraaq/grid-engine/issues/227

currently my workaround is also tile based where I check the position and according to that react based on the tile

**Comment by Annoraaq** [comment]
Hey, so the short answer is that there is currently no other way to detect that. To evaluate on a new feature for that, it would be interesting to know what exactly 'colliding' means in your case. In GridEngine it simply means that a tile position is "blocked" for a character. So it is kept from moving to that position.

So one possible definition of collision in your example could be:

Character A collides with position X if
1. the character is on adjacent position of X
2. the character "tried to move" to position X

"tried to move" would mean that move() was called towards position X.

But there might be more interpretations for that.
Maybe you can describe the exact case and what you dislike about using positionChanged. That might help finding a good solution.

**Comment by Annoraaq** [comment]
I think in a pixel based phyiscs engine the problem is a bit easier. As far as I know it will see something as colliding if it oberlaps or at least "touches".

However in a tile/grid based world overlapping means both objects are on the exact same position and "touching" means you they are at neighboring positions. Both options are probably not what you want, correct?

**Comment by evan-boissonnot** [comment]
Today, I'm trying to add new Layer on my game.
This layer will present all specifics objects I added :

The purpose of these objects : add custom areas to decide when I can detect objects collision.
![image](https://user-images.githubusercontent.com/26603619/156874545-468f15c1-681e-4fd0-accf-a3798f802ee0.png)


I'm not interested about position : I just want when my player collider specifics objects, to emit events, that I can attach from it.

**Comment by zewa666** [comment]
@Annoraaq and @evan-boissonnot I think this now exactly goes in the direction of what we've been talking before with letting the gridEngine "understand" objects and than later on have events for "touching" ( neighbor) and "overlapping" (same tile) which could be two different observables so the user can chose whatever he prefers.

**Comment by evan-boissonnot** [comment]
Hi, 
to be more precise : 

Today, I have tried to use : 

this.map.objects.forEach(element => {
                this.doors = this.map.createFromObjects(element.name, {
                    name: 'door'
                });
});

And with these objects (from object layer), I have tried to add a collision detection, to go to other scene ...
doesn't work :( 

The door sprite (created from Phaser), was not where I have created it on Tiled.

**Comment by Annoraaq** [comment]
I think there are two different problems here:

1. GridEngine objects that have observables you can subscribe. These could tell you whether some character entered a neighboring position or entered their position. That is part of #229 and can already done in a less convenient way by using the positionChanged observables of the characters.

2. Creating sprites from Tiled object layer objects in Phaser. That is something I can't give a lot of advice, because I've never done this. I think there are some Phaser examples handling this: https://labs.phaser.io/view.html?src=src/tilemap/create%20from%20objects.js

Just a clarification for the object layers and GridEngine:
GridEngine does not (and probably will not) support object layers from Tiled because they are - to my knowlege - pixel-based and not tile-based. GridEngine only deals with tile-based positions.

As far as I understood you want to achieve the following:
1. Place objects on pixel positions on an object layer in Tiled
2. Create sprites for those objects with phaser
3. Detect GridEngine characters next to (or on top of) of a tile that contains one of these objects

Step 3 is a bit tricky because you have to define a way to determine the tile position for your object. How you do that depends heavily on the type and shape of your objects. For a square you could get the center and use https://photonstorm.github.io/phaser3-docs/Phaser.Tilemaps.Tilemap.html#worldToTileXY__anchor to get the tile position. Then you could use https://annoraaq.github.io/grid-engine/api/methods.html#positionchangefinished to check whether your character is on the corresponding field. You can also use https://annoraaq.github.io/grid-engine/api/methods.html#getfacingposition to get the position, the character is facing.

**Comment by evan-boissonnot** [comment]
Many thanks @zewa666 and @Annoraaq 

What you say both is the other way I will try. 
As you said, it's the only way with grid engine motor.

But, I don't like the idea to check position.

Collision events are better performance.

**Comment by Annoraaq** [comment]
I don't think that collision events necessarily have better performance. The above `getFacingObjects()` implementation runs in `O(c*o)` where `c` is the number of characters and `o` the number of objects. But you could optimize it by not checking **all** the objects every time. Instead you can keep a hash map that stores all objects for a certain position. Then after every position change of a character you need constant time to check for a facing object (assuming you only need one object per position). That would bring you down to `O(c)`. There are some similarities to this file:  https://github.com/Annoraaq/grid-engine/blob/master/src/GridTilemap/CharBlockCache/CharBlockCache.ts. It is an efficient implementation to check whether a certain position is blocked by a character.

Collision events would basically do the same. On every character position change, the engine has to check whether there is a collision. So I don't see how you can gain any performance advantage by using collision events.

**Comment by zewa666** [comment]
yes very good point. please dont take my example for prod ready. it was merely to depict the idea of how and where to grab data. of course as @Annoraaq suggested you would want to optimize the object retrival plus. plus, if objects are static, one could also memoize lookups for a given facing position and thus even further reduce lookup time.
